File: src/main/java/engine/animation/AnimatedObject.java
package engine.animation;

import com.hackoeur.jglm.Mat4;
import com.hackoeur.jglm.Matrices;
import com.hackoeur.jglm.Vec3;
import com.jogamp.opengl.GL4;
import engine.core.OpenGlObject;
import engine.shader.Shader;

import java.awt.*;
import java.util.ArrayList;
import java.util.Collections;

public class AnimatedObject extends OpenGlObject {

    protected ArrayList<BasicAnimation> animations;
    protected BasicAnimation currentAnim;

    private float frameSizeX;
    private float frameSizeY;

    public AnimatedObject(int bufferParamsCount, int verticesCount, GL4 gl, Dimension boxDim, int id,
                          float frameSizeX, float frameSizeY, BasicAnimation... animationSet) throws Exception {

        super(bufferParamsCount, verticesCount, gl, boxDim, id);

        if (animationSet == null)
            throw new Exception("Must be atleast 1 animation!");

        this.frameSizeX = frameSizeX;
        this.frameSizeY = frameSizeY;
        this.animations = new ArrayList<>();

        Collections.addAll(this.animations, animationSet);
        this.currentAnim = this.animations.get(0);
    }

    public AnimatedObject(int bufferParamsCount, int verticesCount, GL4 gl,
                          float posX, float posY, Dimension boxDim, int id,
                          float frameSizeX, float frameSizeY, BasicAnimation... animationSet) throws Exception {
        super(bufferParamsCount, verticesCount, gl, posX, posY, boxDim, id);

        if (animationSet == null)
            throw new Exception("Must be atleast 1 animation!");

        this.frameSizeX = frameSizeX;
        this.frameSizeY = frameSizeY;
        this.animations = new ArrayList<>();

        Collections.addAll(this.animations, animationSet);
        this.currentAnim = this.animations.get(0);
    }

    @Override
    public void draw(float x, float y, float xSize, float ySize, float rotationAngle, Shader shader) {

        shader.use();

        defineAnimationVariables(shader);

        super.draw(x, y, xSize, ySize, rotationAngle, shader);

    }

    @Override
    public void draw(float xSize, float ySize, float rotationAngle, Shader shader) {

        shader.use();

        defineAnimationVariables(shader);

        super.draw(xSize, ySize, rotationAngle, shader);
    }

    private void defineAnimationVariables(Shader shader) {
        if (this.texture != null || this.textureArray != null) {
            shader.setFloat("xChanging", currentAnim.currentFrameX * frameSizeX, false);
            shader.setInteger("frameNumberX", currentAnim.currentFrameX + 1, false);
            shader.setFloat("yChanging", currentAnim.currentFrameY * frameSizeY, false);
            shader.setInteger("frameY", currentAnim.currentFrameY + 1, false);
        }
    }

    protected void playAnimation() {
        this.currentAnim.changeFrame();
    }

    protected void setAnimation(BasicAnimation a) {
        if (animations.contains(a))
            this.currentAnim = a;
    }

}

File: src/main/java/engine/core/AnimatedObject.kt
package engine.core

import engine.feature.animation.BasicAnimation
import engine.feature.shader.Shader
import engine.feature.shader.ShaderVariableKey

// TODO: what if animations list is empty?
class AnimatedObject(
        var frameSizeX: Float,
        var frameSizeY: Float,
        val animations: MutableList<BasicAnimation>) : Entity {

    var currentAnim: BasicAnimation = animations[0]

    fun defineAnimationVariables(graphicalObject: OpenGlObject2D, shader: Shader) {
        if (graphicalObject.isTextured) {
            shader.setFloat(ShaderVariableKey.Anim.xOffset, currentAnim.currentFrameX * frameSizeX, false)
            shader.setInteger(ShaderVariableKey.Anim.xNumber, currentAnim.currentFrameX + 1, false)
            shader.setFloat(ShaderVariableKey.Anim.yOffset, currentAnim.currentFrameY * frameSizeY, false)
            shader.setInteger(ShaderVariableKey.Anim.yNumber, currentAnim.currentFrameY + 1, false)
        }
    }

    fun playAnimation(deltaTime: Float) {
        currentAnim.play(deltaTime)
    }

    fun setAnimation(a: BasicAnimation) {
        if (animations.contains(a)) {
            currentAnim = a
        }
    }

    override fun toString()= super.toString() + "\n Animated"


}
